cirq/qis/states.py

```diff
@@ -35,6 +36,435 @@
+        actual_qid_shape = (2,) * len(state)
+        if qid_shape is not None and qid_shape != actual_qid_shape:
+            raise ValueError('The specified qid shape must be the same as the '
+                             'qid shape of the given state.\n'
+                             'Specified shape: {qid_shape}\n'
+                             'Shape of state: {actual_qid_state}.')
+        data = state.state_vector().astype(dtype, casting='unsafe')
+        qid_shape = actual_qid_shape
+    elif isinstance(state, int):
+        if qid_shape is None:
+            raise ValueError('The qid shape of the given state is ambiguous. '
+                             'Please specify the qid shape explicitly using '
+                             'the qid_shape argument.')
+        dim = np.prod(qid_shape, dtype=int)
+        data = one_hot(index=state, shape=(dim,), dtype=dtype)
+    else:
+        data = np.array(state)
+        if qid_shape is None:
+            qid_shape = infer_qid_shape(state)
+        if data.ndim == 1:
+            prod = np.prod(qid_shape, dtype=int)
+            if len(qid_shape) == prod and data.dtype.kind != 'c':
+                raise ValueError(
+                    'Because len(qid_shape) == product(qid_shape), it is '
+                    'ambiguous whether the given state contains '
+                    'state vector amplitudes or per-qudit computational basis '
+                    'values. In this situation you are required to pass '
+                    'in a state vector that is a numpy array with a complex '
+                    'dtype.')
+            if data.shape == (len(qid_shape),):
+                # array contains per-qudit computational basis values
+                data = _qudit_values_to_state_tensor(state_vector=data,
+                                                     qid_shape=qid_shape,
+                                                     dtype=dtype)
+    return QuantumState(data=data,
+                        qid_shape=qid_shape,
+                        validate=validate,
+                        dtype=dtype,
+                        atol=atol)
+
+
+def density_matrix(
+        state: np.ndarray,
+        *,  # Force keyword arguments
+        validate: bool = True,
+        dtype: Type[np.number] = np.complex64,
+        atol: float = 1e-7) -> QuantumState:
+    if state.ndim != 2 or len(set(state.shape)) != 1:
+        raise ValueError('A density matrix must be a square matrix. '
+                         f'Got shape {state.shape}.')
+    dim, _ = state.shape
+    qid_shape = _infer_qid_shape_from_dimension(dim)
+    return QuantumState(data=state,
+                        qid_shape=qid_shape,
+                        dtype=dtype,
+                        validate=validate,
+                        atol=atol)
+
+
+def _infer_qid_shape_from_dimension(dim: int) -> Tuple[int, ...]:
+    if dim != 0 and dim & dim - 1 == 0:
+        # dim is a power of 2, assume qubits
```

Comment:
This is a fair assumption, but we should provide a way for users to say that they know their qid_shape already. This would allow users to resolve the ambiguity in your ququad example [here](https://github.com/quantumlib/Cirq/issues/2904#issuecomment-694232713) where a 4x4 density matrix may describe either a state of two qubits or one ququad.

I think we should add qid_shape argument to `density_matrix`. Then we'd call `_infer_qid_shape_from_dimension` only if the user hasn't provided qid_shape explicitly.

Reply:
Done.